core/runtime/v2/shim_manager.go

```diff
@@ -163,37 +164,68 @@ func (m *ShimManager) ID() string {
-			if !ok {
-				return nil, fmt.Errorf("the scheme of sandbox address should be in " +
-					" the form of <protocol>+<unix|vsock|tcp>, i.e. ttrpc+unix or grpc+vsock")
-			}
-			params = shimbinary.BootstrapParams{
-				Version:  int(opts.Version),
-				Protocol: protocol,
-				Address:  address,
+		_, sbErr := m.sandboxStore.Get(ctx, opts.SandboxID)
+		if sbErr != nil {
+			if !errors.Is(sbErr, errdefs.ErrNotFound) {
+				return nil, sbErr
 			}
+
+			log.G(ctx).WithField("id", id).Warningf("sandbox (id=%s) not found, maybe created from v1.x", opts.SandboxID)
+			// NOTE: If sandbox container, like pause, is created by
+			// v1.6.x or v1.7.x, the shim may be not able to group
+			// multiple containers. We should invoke shim binary and
+			// establish new connection based on returned address.
+			shouldInvokeShimBinary = true
 		} else {
-			// For those sandbox we can not get endpoint,
-			// fallback to legacy implementation
-			process, err := m.Get(ctx, opts.SandboxID)
-			if err != nil {
-				return nil, fmt.Errorf("can't find sandbox %s", opts.SandboxID)
-			}
-			p, err := restoreBootstrapParams(process.Bundle())
-			if err != nil {
-				return nil, fmt.Errorf("failed to get bootstrap "+
-					"params of sandbox %s, legacy restore error %v", opts.SandboxID, err)
+			if opts.Address != "" {
+				// The address returned from sandbox controller should
+				// be in the form like ttrpc+unix://<uds-path> or grpc+vsock://<cid>:<port>,
+				// we should get the protocol from the url first.
+				protocol, address, ok := strings.Cut(opts.Address, "+")
+				if !ok {
+					return nil, fmt.Errorf("the scheme of sandbox address should be in " +
+						" the form of <protocol>+<unix|vsock|tcp>, i.e. ttrpc+unix or grpc+vsock")
+				}
+				params = shimbinary.BootstrapParams{
+					Version:  int(opts.Version),
+					Protocol: protocol,
+					Address:  address,
+				}
+			} else {
+				process, err := m.Get(ctx, opts.SandboxID)
+				if err != nil {
+					return nil, fmt.Errorf("can't find shim for sandbox %s: %w", opts.SandboxID, err)
+				}
+
+				p, err := restoreBootstrapParams(process.Bundle())
+				if err != nil {
+					return nil, fmt.Errorf("failed to get bootstrap "+
+						"params of sandbox %s: %w", opts.SandboxID, err)
+				}
+				params = p
 			}
-			params = p
 		}
+	}
+	// Even though one shim can be able to group multiple containers,
+	// it doesn't mean it supports sandbox API. The old shim implementation
+	// still requires containerd to invoke `shim delete` to cleanup
+	// container's resource when each container exits. So, if the
+	// shim version is not higher than 3, we should fallback to invoke
+	// shim binary.
+	//
+	// NOTE: The shim version indicates that the shim supports streaming I/O.
+	// It's rolled out together with the sandbox API and can be used
+	// to determine whether we should invoke the shim binary.
+	const supportSandboxAPIVersion = 3
+	if params.Version < supportSandboxAPIVersion {
```

Comment:
It seems that this should not be here, only when params is valid should it be judged

Reply:
If that param is empty, there are several cases we should invoke shim:
* sandbox(pause) container (there is no sandbox ID yet)
* created container without SandboxID field because it could be created by v1.7. 
* with SandboxID, but the sandbox is not created by podsandbox plugin
* ...

So if param is empty, we should invoke shim.